File: apps/desktop/src/main/defaults.ts
import { join } from "node:path";
import { homedir } from "node:os";
import { generatedDefaults } from "./generated/defaults.ts";

export interface DesktopResolvedDefaults {
  runtime: {
    env: "development" | "production";
  };
  state: {
    rootDir: string;
    logsDir: string;
    runDir: string;
    tuttidDBPath: string;
    tuttidListenerInfoPath: string;
    tuttidLogPath: string;
    desktopLogPath: string;
    tuttidPIDPath: string;
  };
  transport: {
    tcpAddr: string;
  };
  logging: {
    defaultLevel: "debug" | "info" | "warn" | "error";
    defaultOutput: "file" | "stdout" | "tee";
    maxSizeMB: number;
    maxBackups: number;
    maxAgeDays: number;
    maxTotalMB: number;
  };
}

export interface DesktopProtocolClientRegistration {
  scheme: "tutti" | "tutti-dev";
}

export function initializeDesktopEnvironment(options?: {
  appVersion?: string;
  isPackaged?: boolean;
}): void {
  if (!process.env.TUTTI_ENV?.trim()) {
    process.env.TUTTI_ENV = options?.isPackaged ? "production" : "development";
  }

  const appVersion = options?.appVersion?.trim();
  if (!process.env.TUTTI_APP_VERSION?.trim() && appVersion) {
    process.env.TUTTI_APP_VERSION = appVersion;
  }
}

export function resolveDesktopUserDataPath(options: {
  appDataDir: string;
  appName: string;
}): string | null {
  if (resolveTuttiEnv() !== "development") {
    return null;
  }

  const appName = options.appName.trim() || "Tutti";
  return join(options.appDataDir, `${appName}-dev`);
}

export function resolveDesktopDevelopmentAppName(
  appName: string
): string | null {
  if (resolveTuttiEnv() !== "development") {
    return null;
  }

  const safeAppName = appName.trim() || "Tutti";
  return `${safeAppName} Dev`;
}

export function resolveDesktopLoginProtocolScheme(): "tutti" | "tutti-dev" {
  return resolveTuttiEnv() === "development" ? "tutti-dev" : "tutti";
}

export function resolveDesktopLoginCallbackUrl(): string {
  return `${resolveDesktopLoginProtocolScheme()}://login/callback`;
}

export function resolveDesktopLoginProtocolClientRegistration(options: {
  isPackaged: boolean;
}): DesktopProtocolClientRegistration {
  const scheme = resolveDesktopLoginProtocolScheme();
  if (resolveTuttiEnv() === "development" && !options.isPackaged) {
    return { scheme };
  }

  return { scheme };
}

export function resolveDesktopDefaultsFromEnv(): DesktopResolvedDefaults {
  const env = resolveTuttiEnv();
  const stateRootDir = resolveStateRootDir(env);
  const logsDir = resolveLogsDir(stateRootDir);
  const runDir = resolveRunDir(stateRootDir);

  return {
    runtime: {
      env
    },
    state: {
      rootDir: stateRootDir,
      logsDir,
      runDir,
      tuttidDBPath: resolveDBPath(stateRootDir),
      tuttidListenerInfoPath: resolveListenerInfoPath(runDir),
      tuttidLogPath: resolveDaemonLogPath(logsDir),
      desktopLogPath: resolveDesktopLogPath(logsDir),
      tuttidPIDPath: resolvePIDPath(runDir)
    },
    transport: {
      tcpAddr: resolveTCPAddr()
    },
    logging: {
      defaultLevel: generatedDefaults.logging.defaultLevel,
      defaultOutput: generatedDefaults.logging.defaultOutput,
      maxSizeMB: generatedDefaults.logging.maxSizeMB,
      maxBackups: generatedDefaults.logging.maxBackups,
      maxAgeDays: generatedDefaults.logging.maxAgeDays,
      maxTotalMB: generatedDefaults.logging.maxTotalMB
    }
  };
}

export function resolveTuttiEnv(): "development" | "production" {
  return process.env.TUTTI_ENV?.trim().match(/^(dev|development|local)$/i)
    ? "development"
    : "production";
}

function resolveStateRootDir(env: "development" | "production"): string {
  const override = process.env.TUTTI_STATE_DIR?.trim();
  if (override) {
    return override;
  }

  const homeDir = homedir();
  const dirName =
    env === "development"
      ? generatedDefaults.state.developmentDirName
      : generatedDefaults.state.productionDirName;

  return join(homeDir, dirName);
}

function resolveLogsDir(stateRootDir: string): string {
  const override = process.env.TUTTI_LOG_DIR?.trim();
  if (override) {
    return override;
  }

  return join(stateRootDir, generatedDefaults.state.logsDirName);
}
function resolveRunDir(stateRootDir: string): string {
  const override = process.env.TUTTID_RUN_DIR?.trim();
  if (override) {
    return override;
  }

  return join(stateRootDir, generatedDefaults.state.runDirName);
}

function resolveDBPath(stateRootDir: string): string {
  const override = process.env.TUTTID_DB_PATH?.trim();
  if (override) {
    return override;
  }

  return join(stateRootDir, generatedDefaults.state.dbFileName);
}

function resolveDaemonLogPath(logsDir: string): string {
  const override = process.env.TUTTID_LOG_PATH?.trim();
  if (override) {
    return override;
  }

  return join(logsDir, generatedDefaults.state.daemonLogFileName);
}

function resolveDesktopLogPath(logsDir: string): string {
  const override = process.env.TUTTI_DESKTOP_LOG_PATH?.trim();
  if (override) {
    return override;
  }

  return join(logsDir, generatedDefaults.state.desktopLogFileName);
}

function resolvePIDPath(runDir: string): string {
  const override = process.env.TUTTID_PID_PATH?.trim();
  if (override) {
    return override;
  }

  return join(runDir, generatedDefaults.state.pidFileName);
}

function resolveListenerInfoPath(runDir: string): string {
  const override = process.env.TUTTID_LISTENER_INFO_PATH?.trim();
  if (override) {
    return override;
  }

  return join(runDir, generatedDefaults.state.listenerInfoFileName);
}

function resolveTCPAddr(): string {
  return (
    process.env.TUTTID_ADDR?.trim() ||
    generatedDefaults.transport.defaultTCPAddr
  );
}

File: services/tuttid/types/defaults.go
package types

import (
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type generatedDefaultsSpec struct {
	State           generatedStateDefaults
	Transport       generatedTransportDefaults
	Logging         generatedLoggingDefaults
	Analytics       generatedAnalyticsDefaults
	AgentExtensions generatedAgentExtensionDefaults
}

type generatedStateDefaults struct {
	ProductionDirName    string
	DevelopmentDirName   string
	RunDirName           string
	LogsDirName          string
	DBFileName           string
	DaemonLogFileName    string
	DesktopLogFileName   string
	ListenerInfoFileName string
	PIDFileName          string
}

type generatedTransportDefaults struct {
	DefaultTCPAddr string
}

type generatedLoggingDefaults struct {
	DefaultLevel  string
	DefaultOutput string
	MaxSizeMB     int
	MaxBackups    int
	MaxAgeDays    int
	MaxTotalMB    int
}

type generatedAnalyticsDefaults struct {
	AppID         int
	AppKey        string
	Channel       string
	ChannelDomain string
	AppVersion    string
}

type generatedAgentExtensionDefaults struct {
	Sources []generatedAgentExtensionSourceDefaults
}

type generatedAgentExtensionSourceDefaults struct {
	Key              string
	ReleaseIndexURL  string
	SigningKeyID     string
	SigningPublicKey string
	Enabled          bool
}

type AgentExtensionSource struct {
	Key              string
	ReleaseIndexURL  string
	SigningKeyID     string
	SigningPublicKey string
	Enabled          bool
}

type ResolvedDefaults struct {
	Runtime   RuntimeDefaults
	State     StateDefaults
	Transport TransportDefaults
	Logging   LoggingDefaults
}

type RuntimeDefaults struct {
	Env string
}

type StateDefaults struct {
	RootDir                string
	LogsDir                string
	RunDir                 string
	TuttidDBPath           string
	TuttidListenerInfoPath string
	TuttidLogPath          string
	DesktopLogPath         string
	TuttidPIDPath          string
}

type TransportDefaults struct {
	TCPAddr string
}

type LoggingDefaults struct {
	DefaultLevel  string
	DefaultOutput string
	MaxSizeMB     int
	MaxBackups    int
	MaxAgeDays    int
	MaxTotalMB    int
}

type AnalyticsConfig struct {
	Disabled      bool
	Debug         bool
	AppID         int
	AppKey        string
	Channel       string
	ChannelDomain string
	AppVersion    string
}

func ResolveDefaultsFromEnv() ResolvedDefaults {
	env := resolveTuttiEnv()
	stateRootDir := resolveStateRootDir(env)
	logsDir := resolveLogsDir(stateRootDir)
	runDir := resolveRunDir(stateRootDir)

	return ResolvedDefaults{
		Runtime: RuntimeDefaults{
			Env: env,
		},
		State: StateDefaults{
			RootDir:                stateRootDir,
			LogsDir:                logsDir,
			RunDir:                 runDir,
			TuttidDBPath:           resolveDBPath(stateRootDir),
			TuttidListenerInfoPath: resolveListenerInfoPath(runDir),
			TuttidLogPath:          resolveDaemonLogPath(logsDir),
			DesktopLogPath:         resolveDesktopLogPath(logsDir),
			TuttidPIDPath:          resolvePIDPath(runDir),
		},
		Transport: TransportDefaults{
			TCPAddr: resolveTCPAddr(),
		},
		Logging: LoggingDefaults{
			DefaultLevel:  generatedDefaults.Logging.DefaultLevel,
			DefaultOutput: generatedDefaults.Logging.DefaultOutput,
			MaxSizeMB:     generatedDefaults.Logging.MaxSizeMB,
			MaxBackups:    generatedDefaults.Logging.MaxBackups,
			MaxAgeDays:    generatedDefaults.Logging.MaxAgeDays,
			MaxTotalMB:    generatedDefaults.Logging.MaxTotalMB,
		},
	}
}

func ResolveAnalyticsConfig() AnalyticsConfig {
	return AnalyticsConfig{
		Disabled:      resolveAnalyticsDisabled(),
		Debug:         resolveAnalyticsDebug(),
		AppID:         resolveAnalyticsAppID(),
		AppKey:        resolveStringOverride("TUTTI_ANALYTICS_APP_KEY", generatedDefaults.Analytics.AppKey),
		Channel:       generatedDefaults.Analytics.Channel,
		ChannelDomain: resolveStringOverride("TUTTI_ANALYTICS_CHANNEL_DOMAIN", generatedDefaults.Analytics.ChannelDomain),
		AppVersion:    resolveAnalyticsAppVersion(),
	}
}

func ResolveAppVersion() string {
	return resolveStringOverride("TUTTI_APP_VERSION", generatedDefaults.Analytics.AppVersion)
}

func ResolveAgentExtensionSources() []AgentExtensionSource {
	result := make([]AgentExtensionSource, 0, len(generatedDefaults.AgentExtensions.Sources))
	for _, source := range generatedDefaults.AgentExtensions.Sources {
		enabled := source.Enabled
		envName := "TUTTI_AGENT_EXTENSION_" + strings.ToUpper(strings.ReplaceAll(source.Key, "-", "_")) + "_ENABLED"
		if value := strings.TrimSpace(os.Getenv(envName)); value != "" {
			if parsed, err := strconv.ParseBool(value); err == nil {
				enabled = parsed
			}
		}
		result = append(result, AgentExtensionSource{
			Key:              source.Key,
			ReleaseIndexURL:  source.ReleaseIndexURL,
			SigningKeyID:     source.SigningKeyID,
			SigningPublicKey: source.SigningPublicKey,
			Enabled:          enabled,
		})
	}
	return result
}

func resolveTuttiEnv() string {
	value := strings.ToLower(resolveStringOverride("TUTTI_ENV", ""))
	switch value {
	case "dev", "development", "local":
		return "development"
	default:
		return "production"
	}
}

func resolveStateRootDir(env string) string {
	override := resolveStringOverride("TUTTI_STATE_DIR", "")
	if override != "" {
		return override
	}

	homeDir, err := os.UserHomeDir()
	if err != nil || strings.TrimSpace(homeDir) == "" {
		if env == "development" {
			return generatedDefaults.State.DevelopmentDirName
		}
		return generatedDefaults.State.ProductionDirName
	}

	dirName := generatedDefaults.State.ProductionDirName
	if env == "development" {
		dirName = generatedDefaults.State.DevelopmentDirName
	}

	return filepath.Join(homeDir, dirName)
}

func resolveLogsDir(stateRootDir string) string {
	override := resolveStringOverride("TUTTI_LOG_DIR", "")
	if override != "" {
		return override
	}

	return filepath.Join(stateRootDir, generatedDefaults.State.LogsDirName)
}

func resolveRunDir(stateRootDir string) string {
	override := resolveStringOverride("TUTTID_RUN_DIR", "")
	if override != "" {
		return override
	}

	return filepath.Join(stateRootDir, generatedDefaults.State.RunDirName)
}

func resolveDBPath(stateRootDir string) string {
	override := resolveStringOverride("TUTTID_DB_PATH", "")
	if override != "" {
		return override
	}

	return filepath.Join(stateRootDir, generatedDefaults.State.DBFileName)
}

func resolveDaemonLogPath(logsDir string) string {
	override := resolveStringOverride("TUTTID_LOG_PATH", "")
	if override != "" {
		return override
	}

	return filepath.Join(logsDir, generatedDefaults.State.DaemonLogFileName)
}

func resolveDesktopLogPath(logsDir string) string {
	override := resolveStringOverride("TUTTI_DESKTOP_LOG_PATH", "")
	if override != "" {
		return override
	}

	return filepath.Join(logsDir, generatedDefaults.State.DesktopLogFileName)
}

func resolvePIDPath(runDir string) string {
	override := resolveStringOverride("TUTTID_PID_PATH", "")
	if override != "" {
		return override
	}

	return filepath.Join(runDir, generatedDefaults.State.PIDFileName)
}

func resolveListenerInfoPath(runDir string) string {
	override := resolveStringOverride("TUTTID_LISTENER_INFO_PATH", "")
	if override != "" {
		return override
	}

	return filepath.Join(runDir, generatedDefaults.State.ListenerInfoFileName)
}

func resolveTCPAddr() string {
	override := resolveStringOverride("TUTTID_ADDR", "")
	if override != "" {
		return override
	}

	return generatedDefaults.Transport.DefaultTCPAddr
}

func resolveAnalyticsDisabled() bool {
	value := strings.ToLower(resolveStringOverride("TUTTI_ANALYTICS_DISABLED", ""))
	switch value {
	case "":
		return false
	case "1", "true", "yes":
		return true
	case "0", "false", "no":
		return false
	default:
		return true
	}
}

func resolveAnalyticsDebug() bool {
	return resolveTuttiEnv() == "development"
}

func resolveAnalyticsAppID() int {
	override := resolveStringOverride("TUTTI_ANALYTICS_APP_ID", "")
	if override == "" {
		return generatedDefaults.Analytics.AppID
	}

	value, err := strconv.Atoi(override)
	if err != nil || value <= 0 {
		return 0
	}
	return value
}

func resolveAnalyticsAppVersion() string {
	override := resolveStringOverride("TUTTI_ANALYTICS_APP_VERSION", "")
	if override != "" {
		return override
	}
	return ResolveAppVersion()
}

func resolveStringOverride(name string, fallback string) string {
	override := strings.TrimSpace(os.Getenv(name))
	if override != "" {
		return override
	}
	return fallback
}
